frontend/utils/faviconBadge.js

```diff
@@ -0,0 +1,253 @@
+  if (!canvas) {
+    canvas = document.createElement('canvas')
+    canvas.width = size
+    canvas.height = size
+  }
+
+  return canvas
+}
+
+const setFaviconTag = (url) => {
+  if (url) {
+    removeFaviconTag()
+
+    const link = document.createElement('link')
+    link.type = 'image/x-icon'
+    link.rel = 'icon'
+    link.href = url
+    document.getElementsByTagName('head')[0].appendChild(link)
+  }
+}
+
+const drawFavicon = (label, color) => {
+  // fallback to updating the browser title if unsupported
+  if (!(getCanvas() instanceof HTMLCanvasElement) || browser.ie || browser.safari || options.fallback === 'force') {
+    return updateTitle(label)
+  }
+
+  const context = getCanvas().getContext('2d')
+  color = color || '#000000'
+  const src = getCurrentFavicon()
+
+  faviconImage = document.createElement('img')
+  faviconImage.onload = function () {
+    // clear canvas
+    context.clearRect(0, 0, size, size)
+
+    // draw the favicon
+    context.drawImage(faviconImage, 0, 0, faviconImage?.width, faviconImage?.height, 0, 0, size, size)
+
+    // draw bubble over the top
+    if ((label + '').length > 0) drawBubble(context, label, color)
+
+    // refresh tag in page
+    refreshFavicon()
+  }
+
+  // allow cross origin resource requests if the image is not a data:uri
+  // as detailed here: https://github.com/mrdoob/three.js/issues/1305
+  if (!src?.match(/^data/) && options.crossOrigin) {
+    faviconImage.crossOrigin = 'anonymous'
+  }
+
+  faviconImage.src = src
+}
+
+const updateTitle = (label) => {
+  if (options.fallback) {
+    // Grab the current title that we can prefix with the label
+    let originalTitle = document.title
+
+    // Strip out the old label if there is one
+    if (originalTitle[0] === '(') {
+      originalTitle = originalTitle.slice(originalTitle.indexOf(' '))
+    }
+
+    if ((label + '').length > 0) {
+      document.title = '(' + label + ') ' + originalTitle
+    } else {
+      document.title = originalTitle
+    }
+  }
+}
+
+const drawBubble = (context, label, color) => {
+  // automatic abbreviation for long (>2 digits) numbers
+  if (typeof label === 'number' && label > 99 && options.abbreviate) {
+    label = abbreviateNumber(label)
+  }
+
+  // bubble needs to be larger for double digits
+  const len = (String(label)).length - 1
+  const width = options.width * r + (6 * r * len)
+  const height = options.height * r
+  // const top = size - height
+  // const left = size - width - r
+  const bottom = 16 * r
+  const right = 16 * r
+  // const radius = 2 * r
+  const arcRadius = Math.min(width, height) / 2
+
+  // webkit seems to render fonts lighter than firefox
+  context.font = (browser.webkit ? 'bold ' : '') + options.font
+  context.fillStyle = options.background
+  context.strokeStyle = options.background
+  context.lineWidth = r
+
+  // bubble
+  // draw circle bubble
+  context.arc(right - arcRadius, bottom - arcRadius, arcRadius, 0, 2 * Math.PI)
+
+  // draw rectangle bubble
```

Comment:
Can all these lines of code that are commented out be just removed so that `drawBubble()` is more readable?

Reply:
Okay. I will remove all the code which are useless at the moment.

Reply:
Will remove all the lines out of the current requirement. KISS principle!